Interactions + CI/CD + Pagination

NathanealV opened the pull request:
- Converted to using slash commands and interactions
- Using async sqlite library for db interactions
- Updated error handling for cogs
- Added pagination support
- Updated README to include configuration
- Added CI/CD with GitHub Actions and Ansible

### **Backup of database volumes prior to merge**

- [x] Add `CONTRIBUTING.md`

ilikecubesnstuff commented on db.py:
```diff
@@ -1,187 +1,237 @@
+        self.timestamp = timestamp
+
+    def __repr__(self) -> str:
+        return f"Timeout({self.timeout_id}, {self.user}, {self.timestamp})"
```
Where is `self.user` initialized?

ilikecubesnstuff commented on db.py:
```diff
@@ -1,187 +1,237 @@
+import aiosqlite
+from typing import Optional
+
+path = '/database/modmail.db'
```
Maybe capitalize this to indicate it's a constant.

ilikecubesnstuff commented on db.py:
I have a few overall comments for this file.
- Love the use of `contextlib`! Was it not possible to write this context manager using `aiosqlite`'s existing context manager protocols (i.e. using `with` statements)?
- It could be worth abstracting the `db_ops` context management from all the get/set/open/close/update functions via a decorator.
- `Ticket`, `TicketResponse` and `Timeout` look like dataclasses - you should be able to replace the code with dataclass code and have the rest of the code work fine.

NathanealV replied:
- I believe I initially used with statements and existing protocols, however, I found myself repeating the line to set the output of the SQL command to a row format + I believe the existing protocol does not commit/close the connection after each request.
- I wasn't really a fan of making a decorator simply because of the typing system that Python provides but I will look into this
- Dataclasses work much better yeah - will be changed

ilikecubesnstuff commented on db.py:
```diff
@@ -1,187 +1,237 @@
-@database
-def get_ticket(cursor, ticket_id):
+from contextlib import asynccontextmanager
+import aiosqlite
```
Move this import below the standard library imports.

ilikecubesnstuff commented on modmail.py:
```diff
@@ -1,154 +1,70 @@
+        if await db.init():
+            logger.info('Database sucessfully initialized!')
+        else:
+            logger.error('Error while initializing database!')
```
Is this line ever executed? It looks to me as though `db.init()` must always return `True` or throw an exception.

ilikecubesnstuff commented on modmail.py:
```diff
@@ -1,154 +1,70 @@
-
-if success:    
-    bot.run(config['token']) 
+        for cog in ('commands', 'listeners'):
```
Maybe put an explicit list of cogs into a separate variable?

ilikecubesnstuff commented on modmail.py:
```diff
@@ -1,154 +1,70 @@
+                logger.error(e)
+        logger.info("Loaded all cogs.")
+
+        self.add_view(MessageButtonsView(bot, []))
```
Is this for persistent views? Is this to be modified in the future?

NathanealV replied:
Yes it's for persistent views - what do you mean by modified in the future?

ilikecubesnstuff replied:
I thought message IDs were a requirement for persistent views - I guess that shouldn't be the case? Hey, as long as it works 😆

ilikecubesnstuff commented on cogs/listeners.py:
```diff
@@ -1,87 +1,134 @@
+
         response = uformatter.format_message(message)
 
         if not response.strip():
```
Maybe perform `.strip()` inside `uformatter` instead, and then check the truthiness of `response` here?

ilikecubesnstuff commented on utils/actions.py:
```diff
@@ -0,0 +1,294 @@
+
+    if confirmation_view.value is None:
+        return
+    elif confirmation_view.value:
```
Is `if confirmation_view.value` not sufficient?

ilikecubesnstuff commented on utils/actions.py:
```diff
@@ -0,0 +1,294 @@
+    return message
+
+
+async def message_open(bot: commands.Bot, interaction: discord.Interaction,
```
Signature formatting?

ilikecubesnstuff commented on utils/actions.py:
```diff
@@ -0,0 +1,294 @@
+
+    await interaction.response.send_message(embed=close_embed,
+                                            view=confirmation_view)
+    confirmation_view.message = await interaction.original_response()
```
Is it possible to set this attribute during object initialization rather than here?

NathanealV replied:
I initially tried to but couldn't find a way as it's adding an attribute to the `View` itself. I can't really add the message link before the message is sent but I can't send the message without the confirmation_view initialised, unless I send the message first without the view and then edit it onto the message.

ilikecubesnstuff replied:
That does sound finnicky... I was imagining some kind of `confirmation_view.send_message(interaction, embed=close_embed)` that handles sending the message and setting the attribute, but I'm not sure if that improves or worsens complexity/readability. Your call!

ilikecubesnstuff commented on utils/ticket_embed.py:
```diff
@@ -1,7 +1,178 @@
+    @discord.ui.button(emoji="⬅️",
+                       style=discord.ButtonStyle.blurple,
+                       custom_id=f"{config.id_prefix}:before_page")
+    async def before_page(self, interaction: discord.Interaction, _):
```
`previous_page`?

ilikecubesnstuff commented on utils/ticket_embed.py:
```diff
@@ -1,7 +1,178 @@
+            ticket.user) or await interaction.guild.fetch_member(ticket.user)
+        await actions.message_timeout(interaction, member)
+
+    @discord.ui.button(emoji="⬅️",
```
It might be more intuitive if this button appears left-most. I would recommend adding explicit row numbers for the buttons in this file.

NathanealV replied:
I didn't include this button left-most as this wasn't a full pagination view (i.e., first page, previous page, next page, and last page). The pagination buttons appear on the right as a different button colour and are disabled when there is only one page and so I chose this option as the most intuitive.

ilikecubesnstuff commented on utils/ticket_embed.py:
```diff
@@ -143,20 +340,22 @@ def user_timeout(timeout):
     message_embed = discord.Embed(
-        description="You have been timed out. You will be able to message ModMail again after <t:{0}>.".format(timeout)
+        description=
+        f"You have been timed out. You will be able to message ModMail again after <t:{timeout}>."
```
Add `f"... (<t:{timeout}:R>)"` for the end user.

ilikecubesnstuff commented on modmail.py:
```diff
@@ -1,154 +1,70 @@
+            try:
+                await bot.load_extension(f'cogs.{cog}')
+                logger.debug(f'Imported cog "{cog}".')
+            except commands.errors.NoEntryPointError as e:
+                logger.warning(e)
+            except commands.errors.ExtensionNotFound as e:
+                logger.warning(e)
+            except commands.errors.ExtensionFailed as e:
+                logger.error(e)
```
It might be worth exiting entirely if these cogs do not load? Can the bot function at all without them?

I noticed that the setup hooks include system exits, but they might be better placed here since all the components are necessary for this bot from my understanding.

ilikecubesnstuff reviewed (commented):
Quite an overhaul of a PR! I have not looked at the following files:
- `.github/workflows/deploy.yml`
- `.github/workflows/publish.yml`
- `.gitignore`
- `README.md`
- `config.example.json`
- `docker-compose.yml`
- `playbook.yml`
- `requirements.txt`

I also have not done any testing through cloning the repository locally, so my comments are overall about the code formatting/structure and immediate concerns. Overall, I want to comment that:
- Some function signatures have awkward formatting - I'm not sure if there already is a consistent format being followed, but if not, it would be good to decide on one and specify it somewhere.
- Import lists at the top of the file generally are ordered with standard library imports, followed by project dependency imports, followed by local project imports (see [PEP8](https://peps.python.org/pep-0008/#imports)). With the exception of logging imports, I think this is a good standard to follow across all the files (with each group delimited by blank lines).
-  The use of double-quote strings and single-quote strings appears inconsistent - same advice as with function signature formatting.
- Are the slash commands tied to a specific role or channel anywhere (outside the commands cog check)? My concern is whether the slash commands will be visible to a public member in non-modmail channels, regardless of whether they will be able to use it or not.

NathanealV commented:
> * Some function signatures have awkward formatting - I'm not sure if there already is a consistent format being followed, but if not, it would be good to decide on one and specify it somewhere.
> * The use of double-quote strings and single-quote strings appears inconsistent - same advice as with function signature formatting.
> * Import lists at the top of the file generally are ordered with standard library imports, followed by project dependency imports, followed by local project imports (see [PEP8](https://peps.python.org/pep-0008/#imports)). With the exception of logging imports, I think this is a good standard to follow across all the files (with each group delimited by blank lines).

I have switched to using Black as my formatter and so these awkward function formatting (plus other formatting) should be now fixed, or at least standardised.

> * Are the slash commands tied to a specific role or channel anywhere (outside the commands cog check)? My concern is whether the slash commands will be visible to a public member in non-modmail channels, regardless of whether they will be able to use it or not.

Permissions for specific roles and channels can be modified through the dashboard present within the server settings and so should not visible to public members to use.

ilikecubesnstuff commented on pyproject.toml:
What's the `pyproject.toml`'s purpose?

NathanealV replied:
`pyproject.toml` is in preparation for adding versioning and integrating with releases on GitHub.

ilikecubesnstuff commented on .flake8:
Before anything, I just wanted to say [holy mother of jesus](https://github.com/PyCQA/flake8/issues/234).

But also - is 120 lines necessary? Black works with 88 by default (and I don't see any black configuration in your `pyproject.toml` that would suggest using something else here). I would recommend following [this](https://black.readthedocs.io/en/latest/guides/using_black_with_other_tools.html#flake8).

ilikecubesnstuff commented on utils/config.py:
See now that you have a `.toml` file, why not have a `config.toml` and use [tomllib](https://docs.python.org/3/library/tomllib.html)! 😄

On a more serious note, I have a feeling that this file could be made a little cleaner - do you think [dictionary union operations](https://peps.python.org/pep-0584/) with [a copy of `os.environ`](https://stackoverflow.com/questions/13142972/using-copy-deepcopy-on-os-environ-in-python-appears-broken#comment49276709_26074436) could be made use of here? Something like
```py
_env_varnames = {
    "MODMAIL_TOKEN": "token",
    "MODMAIL_APPLICATION_ID": "application_id",
    ...
}
_env = os.environ.copy()
_env_config = {_env_varnames[key]: value for key, value in _env.items() if key in _config}

# or if you're feeling frisky idk
_env_config = {
    confkey: value
    for key, value in _env.items()
    if (confkey := key.replace("MODMAIL_", "").lower()) in _config
}
```
followed by
```py
_config = json.load(open(_path, "r"))
_config |= _env_config
```
followed by
```py
token = _config["token"]
application_id = _config["application_id"]
...
```

NathanealV replied:
The manual configuration and loading logic has been switched out for the use of [ConfZ](https://github.com/Zuehlke/ConfZ) and it's quite easy to get up and running - it has been tested for both JSON and env with Docker.

ilikecubesnstuff reviewed (commented):
Thank you for making the changes! I had not heard of [Black](https://black.readthedocs.io/en/latest/index.html) before and it looks like I've really been missing out.

I've written some additional comments for you. :D

ilikecubesnstuff reviewed (approved):
Looks good! The config implementation with `ConfZ` is very clean, nice work.
